pkg/cortex/modules.go

```diff
@@ -236,6 +231,20 @@ func (t *Cortex) initQuerier(cfg *Config) (err error) {
 	subrouter.Path("/validate_expr").Handler(t.httpAuthMiddleware.Wrap(http.HandlerFunc(t.distributor.ValidateExprHandler)))
 	subrouter.Path("/chunks").Handler(t.httpAuthMiddleware.Wrap(querier.ChunksHandler(queryable)))
 	subrouter.Path("/user_stats").Handler(middleware.AuthenticateUser.Wrap(http.HandlerFunc(t.distributor.UserStatsHandler)))
+
+	// Start the query frontend worker once the query engine and the store
+	// have been successfully initialized.
+	t.worker, err = frontend.NewWorker(cfg.Worker, httpgrpc_server.NewServer(t.server.HTTPServer.Handler), util.Logger)
+	if err != nil {
+		return
+	}
+
+	// Once the execution reaches this point, all synchronous initialization has been
+	// done and the querier is ready to serve queries, so we're just returning a 202.
+	t.server.HTTP.Path("/ready").Handler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
`nit` I think I'd rather see it return something like a 503 before its ready, instead of timing out. And then switch to the 204 once it gets here. But I don't think it's a big deal

Reply:
Yeah. The reason why I didn't do it, is because the HTTP server is not opened until the `cortex.Run()` is executed, which occurs after all the modules have been initialized.

The cleaner solution would be not having any synchronous initialization in the modules initialization (so that modules are fast to initialize), while making sure we do start the querier worker only after the asynchronous initial sync is completed, but this would require some significative refactoring which I'm not sure it's worth at this stage.

Thoughts?

Reply:
Definitely not worth it right now. This change is good